### PR TITLE
feat(controller): account co-located fork VMs against the source pod memory budget

### PR DESCRIPTION
**Pull request by stubbi**
## Thinking Path

> - Mitos boots Firecracker microVMs and forks them via copy-on-write snapshots, exposed through CRDs; the multi-VM-per-pod husk work runs a tenant's fork swarm as many same-tenant VMs inside one session pod.
> - This touches the controller: the MultiVMFork routing in internal/controller/sandboxfork_controller.go, which decides whether a fork child becomes an ADDITIONAL VM spawned inside the source pod or a brand-new child pod.
> - L1.7a shipped that routing behind the --multi-vm-fork flag but gated co-location at a hardcoded maxCoLocatedForkVMsPerPod = 4, a conservative stand-in with no relationship to how much memory the pod can actually hold; stacking N full guests into a pod sized for one can push it past its cgroup memory.max and a sibling OOM-kills a sibling (the user's own live work).
> - Guarantee A of the multinode plan says a fork must never overcommit a pod: account each added VM against a per-pod memory budget, safe at the CoW worst case, and pend or spill when the pod is full. Memory is non-compressible, so this is a hard requirement, not a soft one.
> - This pull request replaces the hardcoded count with a real per-pod memory budget: a fork co-locates a child only while floor(pod memory.max / per-VM guest RAM) minus the source VM's own slot has room, reserving each co-located VM its FULL guest memory (worst case, not the CoW-typical dirty set), and spills the rest to a new honestly node-scheduled pod.
> - The benefit is that a fork can never overcommit a session pod's memory: co-location is bounded by what the pod can actually hold, so no sibling can OOM a sibling, and a pod sized for one VM co-locates nothing and spills every child.

## Linked Issues or Issue Description

No tracked issue. Increment L1.7b of the multi-VM-per-pod husk work, implementing the per-pod dimension of guarantee A (a fork never overcommits a node's memory) from docs/superpowers/plans/2026-07-06-fork-primitive-multinode.md. Problem: L1.7a co-located fork VMs into the source pod up to a hardcoded count with no memory accounting, so a fork could pack more guest RAM into a pod than its cgroup memory limit holds, letting a sibling VM OOM-kill a sibling under worst-case dirtying.

## What Changed

- Added `coLocatedForkVMBudget` in internal/controller/sandboxfork_controller.go: derives the co-location budget from the source pod's actual husk-container memory accounting as `floor(Limits[memory] / Requests[memory]) - 1` (the minus one reserves the source VM already resident in the pod), floored at 0. A malformed or resource-free pod yields 0 (co-locate nothing, spill everything).
- Removed the hardcoded `maxCoLocatedForkVMsPerPod = 4` const and its test-only export `MaxCoLocatedForkVMsPerPod`.
- The MultiVMFork routing now co-locates a child only while `int(i) < coLocationBudget` (computed once per fan-out); slots past the budget fall through to the byte-for-byte-unchanged new-pod spill path.
- Updated the routing comment in sandboxclaim_controller.go to describe the memory budget rather than the fixed cap.
- Tests: reworked TestMultiVMForkSpillsPastPodBudgetToNewPods to drive a precise pod budget (1024Mi limit / 256Mi per VM = 4 total, 3 co-locate, 2 spill); added TestMultiVMForkPodBudgetHoldsOneVMSpills (768Mi / 512Mi = 1 total, 0 co-locate, the one child spills and spawn-vm is never called); added a `withCoLocationBudget` pod mutator and stamped a budget onto the co-location-expecting tests. The flag-off, non-multi-vm-source, spawn-error, and flag-flip-off tests are unchanged in intent and still green.
- Docs: added a Status note to the guarantee A section of the multinode plan describing the landed per-pod admit-and-spill accounting and the deferred node-level refinement.

## Verification

- [x] `go build ./...`: clean (after `go clean -cache` to clear a corrupted local build cache; no code errors).
- [x] `go test ./internal/controller/...` (envtest 1.31 via `~/go/bin/setup-envtest use 1.31`): `ok  mitos.run/mitos/internal/controller  239.522s`. MultiVMFork suite verbose, all 7 pass including TestMultiVMForkSpillsPastPodBudgetToNewPods and TestMultiVMForkPodBudgetHoldsOneVMSpills. The two new tests were confirmed RED against the hardcoded-cap implementation first (TDD), then GREEN after the budget landed.
- [x] `golangci-lint run --timeout=5m`: only the pre-existing darwin-only `internal/fork/uffd.go:42:22: func uffdMapping.pageSizeBytes is unused` finding, nothing new.
- [x] `GOOS=linux golangci-lint run --timeout=5m`: clean, no findings.

## Risks

Low risk. The change is confined to the co-location admission decision of the MultiVMFork routing, which is behind the --multi-vm-fork controller flag (default off); the flag-off path and the non-multi-vm-source fallback are byte-for-byte the existing new-pod path, so nothing changes for the default deployment. When the flag is on, the new budget is strictly more conservative than the old fixed cap under the honest default pod sizing (it co-locates fewer VMs, never more), so the failure mode is spilling a fork to a new pod rather than overcommitting one. This increment lands only the per-pod dimension; a dedicated node-budget check for the co-located dimension and the in-place-pod-resize option are deferred to a follow-up (the spill path already keeps the node honest via the existing scheduler and capacity-admission gate).

## Model Used

Claude Opus (Anthropic), model id claude-opus-4-8, extended thinking enabled. The agent wrote the failing tests first, implemented the budget, and ran the build, envtest suite, and both golangci-lint invocations.

## Checklist

- [x] PR title is a conventional commit (feat, fix, docs, ci, chore, refactor, test)
- [x] Thinking Path traces from project context to this change
- [x] Model Used is filled in (with version and capability details)
- [x] Tests added for behavior changes, in the same commit (TDD)
- [x] Docs updated in the same PR
- [ ] Threat-model delta not applicable: the security surface (isolation boundary, egress, secrets) is unchanged; this only tightens memory admission.
- [ ] Benchmark run not applicable: no new public number; the change makes an O(1) admission decision at fork time and states no benchmark claim.
- [x] No em or en dashes introduced anywhere
- [x] Secret values never logged, in errors, in condition messages, or on host paths
- [x] No internal/instance-local references (only public `#NNN` / mitos-run/mitos URLs)
- [x] Every commit carries a `Signed-off-by` trailer (`git commit -s`)

## Summary by CodeRabbit

* **Bug Fixes**
  * Multi-VM forks now decide source-pod co-location vs spilling using the source pod’s memory request/limit budget, replacing the prior fixed co-location cap.
  * Added edge-case handling so when only the source VM fits the budget, child VMs reliably spill without attempting in-pod co-location.
* **Tests**
  * Updated and expanded envtest coverage to make co-location vs spill behavior deterministic and to validate non-wedging spawn behavior.
* **Documentation**
  * Clarified the “fork never overcommits memory” guarantee with the admit-and-spill approach and budget-based routing.

### CODE REVIEW

File: internal/controller/sandboxfork_controller.go [hunk 1]
```diff
@@ -35,15 +36,63 @@ type huskForkSnapshotRemover func(ctx context.Context, addr string, tlsConf *tls
...
+	if !found || perVM.IsZero() || limit.IsZero() {
+		return 0
+	}
+	total := limit.Value() / perVM.Value()
+	coLocated := total - 1
+	if coLocated < 0 {
+		return 0
+	}
+	return int(coLocated)
+}
```
[the remark this answers is not in the document]
Comment (stubbi) [reply internal/controller/sandboxfork_controller.go:110]: Valid finding, and now documented precisely. Two points: (1) the NODE is not overcommitted regardless: the pod's cgroup memory LIMIT is a hard ceiling the node scheduler already reserved, so an over-admission from concurrent same-source forks is caught INSIDE the pod (a co-located guest exceeding memory.max is OOM-killed, fail-closed at the pod boundary), never as a node overcommit. Guarantee A's node dimension holds via the pod limit. (2) What is genuinely missing, as you say, is a cross-fork SHARED RESERVATION so an over-budget co-located child cleanly SPILLS to a new pod instead of risking that intra-pod OOM under concurrent forks. A race-free occupancy count across forks is a heavy lift (TOCTOU between concurrent reconciles needs a reservation with conflict retry), so it is a tracked follow-up (L1.7c); the flag is default off and the L1.8 canary is single-fork, so nothing ships that claims cross-fork enforcement. I have corrected the coLocatedForkVMBudget doc to state the per-fork scope + the node-protected-by-pod-limit reasoning + the follow-up explicitly, rather than overclaiming full enforcement.
